src/Microsoft.ML.Data/Utilities/LocalEnvironment.cs

```diff
@@ -42,9 +45,11 @@ protected override void DisposeCore()
         /// </summary>
         /// <param name="seed">Random seed. Set to <c>null</c> for a non-deterministic environment.</param>
         /// <param name="conc">Concurrency level. Set to 1 to run single-threaded. Set to 0 to pick automatically.</param>
-        public LocalEnvironment(int? seed = null, int conc = 0)
+        /// <param name="compositionContainerFactory">The function to retrieve the composition container</param>
+        public LocalEnvironment(int? seed = null, int conc = 0, Func<CompositionContainer> compositionContainerFactory = null)
             : base(RandomUtils.Create(seed), verbose: false, conc)
         {
+            _compositionContainerFactory = compositionContainerFactory;
```

Comment:
>_compositionContainerFactory = compositionContainerFactory [](start = 12, length = 58)

So, why a *factory*? Is there something important about deferred execution of the method, or some reason I have it invoked multiple times?

Reply:
1. It is invoked multiple times: the user may have added more parts to the `PartCatalog`, so the old composition containers are no longer valid.
2. My first approach used to expose the catalog itself: this didn't work because I need to explicitly register *a particular instance* of `MLContext` in the container.

---
In reply to: [229093844](https://github.com/dotnet/machinelearning/pull/1406#discussion_r229093844) [](ancestors = 229093844)

Reply:
Oh gosh. Well I guess that is invisible to the user, I suppose

---
In reply to: [229396906](https://github.com/dotnet/machinelearning/pull/1406#discussion_r229396906) [](ancestors = 229396906,229093844)